### PR TITLE
[CBRD-24855] Memory Monitoring - Register heap file module

### PR DESCRIPTION
**Pull request by Rudeus**
http://jira.cubrid.org/browse/CBRD-24855

Register heap module
 - When base data structure is returned to its free-list, free-area, etc.,  subtract transaction memory usage.
 - Add mmon_tracking_tag in THREAD_ENTRY
 - Add memory tracking functions at base data structures
 - Add memory tracking functions at heap file module
   - will be upload per component

### CODE REVIEW

**Comment by Rudeus** [comment]
The content provided so far is related to 'Add mmon_tracking_tag in THREAD_ENTRY' and 'Ready for adding memory tracking function like adding module_id, stat_id, etc.'.

**Comment by Rudeus** [comment]
The content provided so far is related to 'Add mmon_tracking_tag in THREAD_ENTRY' and 'Ready for adding memory tracking function like adding module_id, stat_id, etc.'.

File: src/base/memory_monitor_sr.cpp [hunk 1]
```diff
@@ -588,6 +599,15 @@ namespace cubperf
   {
     add_stat (src, - (size));
     add_stat (dest, size);
+
+    /* If move is occurred from other modules to MMON_COMMON,
+     * it means that memory is returned to common structures for keeping.
+     * ex) pool, free list, reserved, etc..
+     * In this case, transaction's memory usage will be subtracted. */
+    if (dest == MMON_COMMON)
```
Comment (joohok) [inline src/base/memory_monitor_sr.cpp]: The `if (src == MMON_COMMON)` condition will be addressed in following commit. 
it is discussed offline

**Comment by Rudeus** [comment]
for tracking base data structures, add sub/move/resize functions for using `tracking_tag`.

**Comment by Rudeus** [comment]
Add memory tracking functions for tracking memory usage of base data structures:
 - add memory tracking functions
 - add fake functions lock-free unit test

File: src/base/memory_monitor_sr.cpp [hunk 2]
```diff
@@ -720,3 +731,54 @@ void mmon_aggregate_tran_info (int tran_count, MMON_TRAN_INFO &info)
 
   mmon_Gl->aggregate_tran_info (tran_count, info);
 }
+
+MMON_STAT_ID mmon_set_tracking_tag (MMON_STAT_ID new_tag)
+{
+  MMON_STAT_ID prev_tag;
+  THREAD_ENTRY *cur_thread_p;
+
+  cur_thread_p = thread_get_thread_entry_info ();
```
Comment (wrlawodms) [inline src/base/memory_monitor_sr.cpp]: What if `cur_thread_p` is `nullptr`?  Isn't a category needed for this like UNKNOWN? What's the design for this case?

File: src/base/memory_monitor_cl.cpp [hunk 3]
```diff
@@ -152,7 +152,14 @@ mmon_print_tran_info (MMON_TRAN_INFO &tran_info)
 
   for (const auto &t_stat : tran_info.tran_stat)
     {
-      fprintf (stdout, "\t%14d | %17lu\n", t_stat.tranid, t_stat.cur_stat);
+      if (t_stat.cur_stat >= 0)
+	{
+	  fprintf (stdout, "\t%14d | %17lu\n", t_stat.tranid, t_stat.cur_stat);
+	}
+      else
+	{
```
Comment (wrlawodms) [inline src/base/memory_monitor_cl.cpp:181]: It's a non-self-describe code that needs a comment.

File: src/base/memory_monitor_sr.cpp [hunk 4]
```diff
@@ -720,3 +731,54 @@ void mmon_aggregate_tran_info (int tran_count, MMON_TRAN_INFO &info)
...
+  mmon_add_stat (cur_thread_p, cur_thread_p->mmon_tracking_tag, size);
+}
+
+void mmon_sub_stat_with_tracking_tag (int64_t size)
+{
+  THREAD_ENTRY *cur_thread_p = thread_get_thread_entry_info ();
+  mmon_sub_stat (cur_thread_p, cur_thread_p->mmon_tracking_tag, size);
+}
+
+void mmon_move_stat_with_tracking_tag (int64_t size, bool tag_is_src)
```
Comment (wrlawodms) [inline src/base/memory_monitor_sr.cpp]: I think you'd better off dividing these into two `mmon_move_stat_to_common_from_tracking_tag` and `mmon_move_stat_to_tracking_tag_from_common`. Or, whatever it is,  the name should include that it is moved to `MMON_COMMON`.

File: src/base/object_representation_sr.c [hunk 5]
```diff
@@ -1634,6 +1652,9 @@ or_install_btids_foreign_key_ref (DB_SEQ * fk_container, OR_INDEX * index)
 
       if (set_get_element_nocopy (fk_seq, 1, &val) != NO_ERROR)
 	{
+#ifdef SERVER_MODE
+	  mmon_sub_stat_with_tracking_tag (sizeof (OR_FOREIGN_KEY));
```
Comment (wrlawodms) [inline src/base/object_representation_sr.c:1674]: Why not subtract above two error cases, L1646 and L1638?

File: src/base/object_representation_sr.c [hunk 6]
```diff
@@ -1894,13 +1964,20 @@ or_install_btids_class (OR_CLASSREP * rep, BTID * id, DB_SEQ * constraint_seq, i
       er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (OR_ATTRIBUTE *) * att_cnt);
       return;
     }
+#ifdef SERVER_MODE
+  index->att_cnt = att_cnt;
```
Comment (wrlawodms) [inline src/base/object_representation_sr.c:1986]: Is it fixing a missing part? seems irrelevant to this issue.

**Review by wrlawodms — COMMENTED** [review]
IN progress..

**Comment by Rudeus** [comment]
Add memory tracking functions for heap_file module.

**Comment by Rudeus** [comment]
finish local test

File: msg/en_US.utf8/utils.msg [hunk 7]
```diff
@@ -1330,4 +1330,6 @@ valid options:\n\
                                   WARNING: This option has to be used with '-t(--transaction)' option.\n\
     -m, --module=TYPE             display detailed memory usage information of the specified TYPE.\n\
                                   TYPE of modules\n\
-                                  -1      all modules\n
+                                  -1      all modules\n\
+                                   0      heap module\n\
+                                   1      common module\n
```
Comment (wrlawodms) [inline msg/en_US.utf8/utils.msg]: I'd put `common module` before `heap module`. If other modules come in, it will looks strange. It'd make `heap module` special.

File: src/base/area_alloc.c [hunk 8]
```diff
@@ -149,14 +160,21 @@ area_create (const char *name, size_t element_size, size_t alloc_count)
   size_t adjust;
 #if defined (SERVER_MODE)
   int rv;
+  MMON_STAT_ID prev_tag = mmon_set_tracking_tag (MMON_COMMON);
```
Comment (wrlawodms) [inline src/base/area_alloc.c:163]: You can make `(void) mmon_set_tracking_tag (prev_tag);` called automatically and safely with an RAII object. It doesn't only look good, but it's also safe. Please consider this. 
Example:
```
class mmon_tag_tracker
{
        public:
        mmon_tag_tracker (int tag)
        {
                cur_tag = tag;
                prev_tag = mmon_set_tracking_tag (tag);
        }

        ~mmon_tag_tracker ()
        {
                int ttt =  mmon_set_tracking_tag (prev_tag);
                assert (origin_tag == ttt);
        }

        private:
                int prev_tag;
                int origin_tag; 
}

```
```suggestion
  mmon_tag_tracker tracker track (MMON_COMMON);
```

File: src/base/memory_monitor_sr.cpp [hunk 9]
```diff
@@ -30,6 +30,11 @@
 #include "memory_monitor_sr.hpp"
 #include "log_impl.h"
 
+#if !defined (NDEBUG)
+thread_local char mmon_err_msg[DB_MAX_IDENTIFIER_LENGTH];
```
Comment (wrlawodms) [inline src/base/memory_monitor_sr.cpp]: Please elaborate on why this must be thread_local.

File: src/base/memory_monitor_sr.cpp [hunk 10]
```diff
@@ -618,6 +667,61 @@ namespace cubperf
   {
     m_aggregater.get_transaction_info (tran_count, info);
   }
+
+  void memory_monitor::stat_dump_debug ()
```
Comment (wrlawodms) [inline src/base/memory_monitor_sr.cpp]: It's not about heap module. Please separate it to another issue.

File: src/base/memory_monitor_sr.cpp [hunk 11]
```diff
@@ -720,3 +831,81 @@ void mmon_aggregate_tran_info (int tran_count, MMON_TRAN_INFO &info)
...
 }
+
+MMON_STAT_ID mmon_set_tracking_tag (MMON_STAT_ID new_tag)
+{
+  MMON_STAT_ID prev_tag = MMON_STAT_LAST;
+  THREAD_ENTRY *cur_thread_p;
+
+  if (mmon_Gl != nullptr)
+    {
+      cur_thread_p = thread_get_thread_entry_info ();
```
Comment (wrlawodms) [inline src/base/memory_monitor_sr.cpp:757]: What happens `cur_thread_p` is nullptr? It can be null though I'm not sure current usages run into this.

File: src/compat/db_json.cpp [hunk 12]
```diff
@@ -856,6 +871,12 @@ JSON_VALIDATOR::JSON_VALIDATOR (const JSON_VALIDATOR &copy)
   else
     {
       m_schema_raw = strdup (copy.m_schema_raw);
+#ifdef SERVER_MODE
+      if (m_schema_raw != NULL)
```
Comment (wrlawodms) [inline src/compat/db_json.cpp:875]: Do we need to consider this kind of exception while the existing code doesn't do it?

File: src/base/memory_monitor_sr.cpp (hunk 10 above)
Comment (hornetmj) [reply src/base/memory_monitor_sr.cpp]: I agree with him. I would be nice to handle another PR or issue.

**Review by hornetmj — APPROVED** [review]

File: src/base/memory_monitor_cl.cpp [hunk 13]
```diff
@@ -126,9 +131,15 @@ mmon_print_module_info_summary (uint64_t server_mem_usage, std::vector<MMON_MODU
 
   for (const auto &m_info : module_info)
     {
+      // TODO: It's a temporary measure for hiding MMON_OTHERS
+      if (!strcmp (m_info.name, "OTHERS"))
```
Comment (joohok) [inline src/base/memory_monitor_cl.cpp:137]: will this part be removed? 
I think It is unclear what the TODO is and when it should be done

File: src/base/memory_monitor_common.h [hunk 14]
```diff
@@ -29,19 +29,19 @@
 
 typedef enum mmon_module_id
 {
-  /* TODO: this dummy modules will be changed when heap module is registered */
   MMON_MODULE_ALL = -1,
-  MMON_MODULE_DUMMY,
-  MMON_MODULE_LONG_DUMMY,
+  MMON_MODULE_HEAP,
+  MMON_MODULE_COMMON,
+  MMON_MODULE_OTHERS,
```
Comment (joohok) [inline src/base/memory_monitor_common.h]:
```suggestion
  MMON_MODULE_COMMON,
  MMON_MODULE_HEAP,
  /* Add new modules before this */
  MMON_MODULE_OTHERS,
```

Maybe same opinion with above (in utils.msg)
And it would be better to leave comments on where to add new modules.

File: src/object/elo.c [hunk 15]
```diff
@@ -288,6 +318,11 @@ elo_copy (DB_ELO * elo, DB_ELO * dest)
 		ret = er_errid ();
 		goto error_return;
 	      }
+#ifdef SERVER_MODE
+	    mmon_add_stat_with_tracking_tag (strlen (elo->locator) + 1);
+	    strcpy (out_uri, elo->locator);
```
Comment (joohok) [inline src/object/elo.c]: why is this string copy needed?

**Review by joohok — COMMENTED** [review]
not finished yet

File: msg/en_US.utf8/utils.msg (hunk 7 above)
Comment (Rudeus) [reply msg/en_US.utf8/utils.msg]: Done.

File: src/base/memory_monitor_sr.cpp (hunk 2 above)
Comment (Rudeus) [reply src/base/memory_monitor_sr.cpp]: We do not assume that `thread_p == NULL`. Since `mmon_Gl` is initialized after the initialization of the `thread_manager`, it guarantees accessibility to all threads managed by the `thread_manager`. If another developer is creating threads without using `THREAD_ENTRY`, the issue is whether these threads should be incorporated into the thread_manager. Considering this should be a separate concern.

File: src/base/memory_monitor_cl.cpp (hunk 3 above)
Comment (Rudeus) [reply src/base/memory_monitor_cl.cpp:181]: Done.

File: src/base/memory_monitor_sr.cpp (hunk 4 above)
Comment (Rudeus) [reply src/base/memory_monitor_sr.cpp]: Done. I think that it is no use deviding MMON_COMMON use case, just the tag is used for source or destination.

File: src/base/object_representation_sr.c (hunk 5 above)
Comment (Rudeus) [reply src/base/object_representation_sr.c:1674]: Done.

File: src/base/object_representation_sr.c (hunk 6 above)
Comment (Rudeus) [reply src/base/object_representation_sr.c:1986]: It is for getting size of the structures when they are freed.

File: src/base/area_alloc.c (hunk 8 above)
Comment (Rudeus) [reply src/base/area_alloc.c:163]: It can be added c-only code.

File: src/base/memory_monitor_sr.cpp (hunk 9 above)
Comment (Rudeus) [reply src/base/memory_monitor_sr.cpp]: I devided it to another issue.

File: src/base/memory_monitor_sr.cpp (hunk 10 above)
Comment (Rudeus) [reply src/base/memory_monitor_sr.cpp]: I devided it to another issue.

File: src/base/memory_monitor_sr.cpp (hunk 11 above)
Comment (Rudeus) [reply src/base/memory_monitor_sr.cpp:757]: We do not assume that thread_p == NULL. Since mmon_Gl is initialized after the initialization of the thread_manager, it guarantees accessibility to all threads managed by the thread_manager. If another developer is creating threads without using THREAD_ENTRY, the issue is whether these threads should be incorporated into the thread_manager. Considering this should be a separate concern.

File: src/compat/db_json.cpp (hunk 12 above)
Comment (Rudeus) [reply src/compat/db_json.cpp:875]: In the reference of `strdup()`, it can return 0 when it fails memory allocation.

File: src/base/memory_monitor_cl.cpp (hunk 13 above)
Comment (Rudeus) [reply src/base/memory_monitor_cl.cpp:137]: Done.

File: src/base/memory_monitor_common.h (hunk 14 above)
Comment (Rudeus) [reply src/base/memory_monitor_common.h]: Done.

File: src/object/elo.c (hunk 15 above)
Comment (Rudeus) [reply src/object/elo.c]: Because if the function goes to fail, it has to free allocated memories, but we don't know which case is hit in this circumstance. So I have to make recognize all string allocations at once. `out_uri` is a static char array.

File: src/base/area_alloc.c (hunk 8 above)
Comment (wrlawodms) [reply src/base/area_alloc.c:163]: A c-complier can compile this code, can't it? It's just a declaration of a structure in the respective of c-compiler.

File: src/base/memory_monitor_cl.cpp (hunk 3 above)
Comment (wrlawodms) [reply src/base/memory_monitor_cl.cpp:181]: I don't understand the comment. What makes it negative?

File: src/base/memory_monitor_sr.cpp (hunk 11 above)
Comment (wrlawodms) [reply src/base/memory_monitor_sr.cpp:757]: OK, let's assume that only threads managed by the thread_manager use the mmon module.

File: src/base/object_representation_sr.c (hunk 6 above)
Comment (wrlawodms) [reply src/base/object_representation_sr.c:1986]: Why should `index->att_cnt` be set to "get size of the structure"?

File: src/compat/db_json.cpp (hunk 12 above)
Comment (wrlawodms) [reply src/compat/db_json.cpp:875]: I know it can return 0. What I meant is that if the `strdup` fails, the existing code doesn't seem to work. It means this code doesn't assume it can be 0. It's not for this code, but in general.
Or, is this allowed to be NULL? or just for safe?

**Review by wrlawodms — APPROVED** [review]

File: src/object/elo.c [hunk 16]
```diff
@@ -338,11 +376,17 @@ elo_copy (DB_ELO * elo, DB_ELO * dest)
 error_return:
   if (locator != NULL)
     {
+#ifdef SERVER_MODE
+      mmon_add_stat_with_tracking_tag (strlen (out_uri) + 1);
```
Comment (joohok) [inline src/object/elo.c]:
```suggestion
      mmon_sub_stat_with_tracking_tag (strlen (out_uri) + 1);
```

File: src/object/elo.c (hunk 15 above)
Comment (joohok) [reply src/object/elo.c]: If the length of the string is required when the it fails, then we don't need to copy the string I think. Instead we can store the length somewhere

File: src/object/elo.c (hunk 16 above)
Comment (Rudeus) [reply src/object/elo.c]: Done.

File: src/object/elo.c (hunk 15 above)
Comment (Rudeus) [reply src/object/elo.c]: I checked it and I agree that we don't need to use `out_uri` and strcpy. Done.

**Review by joohok — APPROVED** [review]
